ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Query.hs

```diff
@@ -76,6 +85,31 @@ getLedgerDB ::
   => ChainDbEnv m blk -> STM m (LgrDB.LedgerDB' blk)
 getLedgerDB CDB{..} = LgrDB.getCurrent cdbLgrDB
 
+-- | Get a 'HeaderStateHistory' populated with the 'HeaderState's of the
+-- last @k@ blocks of the current chain.
+getHeaderStateHistory ::
+     forall m blk.
+     ( IOLike m
+     , HasHardForkHistory blk
+     , HasAnnTip blk
+     , IsLedger (LedgerState blk)
+     )
+  => ChainDbEnv m blk -> STM m (HeaderStateHistory blk)
+getHeaderStateHistory cdb@CDB{cdbTopLevelConfig = cfg} = do
+    ledgerDb <- getLedgerDB cdb
+    let currentLedgerState = ledgerState $ LedgerDB.ledgerDbCurrent ledgerDb
+        -- This summary can convert all tip slots of the ledger states in the
```

Comment:
Unless there's a Chain Growth violation, right?

That's probably fine for now --- lots goes wrong with CG is violated.

But ideally the ChainSel logic could be calculating these incrementally, such that this function is merely a lookup.

Reply:
This should work even in case of a chain growth violations. Note that we use the ledger state *at the tip* of the LedgerDB in order to translate the time of tip slots of all previous ledger states.
![2024-09-04_17-07](https://github.com/user-attachments/assets/472c531f-5f8e-41e0-87b3-4459ca5b5081)

Reply:
Yep, I had it backwards. I think I missed the "not" in the "not newer" you wrote.